Add implementation of LyCORIS LoKr (KronA-like adapter) for SD&SDXL models

kovalexal opened the pull request:
This PR focuses on increasing compatibility of SD&SDXL adapters in peft with other open-source instruments like [LyCORIS](https://github.com/KohakuBlueleaf/LyCORIS). Feel free to learn more about LyCORIS adapters from resources like [this](https://stable-diffusion-art.com/lycoris/).

This specific PR is currently aimed at adding proper compatibility of PEFT with LoKr adapters. LoKr adapters are described in detail in [LyCORIS](https://arxiv.org/abs/2309.14859) paper and are inspired by [KronA](https://arxiv.org/abs/2212.10650) paper.

Currently, it's a draft PR, so the following things are needed:

- [x] Actual implementation of LoKr
- [x] Unit tests
- [x] Conversion script for SD&SDXL for [kohya_ss](https://github.com/bmaltais/kohya_ss) / [LyCORIS](https://github.com/KohakuBlueleaf/LyCORIS) trained LoKrs
- [x] Sample training script for SD / SDXL / SD&SDXL

It would be great to take into account that LoKr may reuse a lot of code from LoHa #956.

@pacman100 @BenjaminBossan @younesbelkada FYI

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_978). All of your documentation changes will be reflected on that endpoint.

BenjaminBossan commented:
Not quite sure what the status is, is the PR ready for review or are you still working on it? Let us know when it is ready.

kovalexal commented:
@BenjaminBossan hi!

I've just finished LoKr implementation. As expected, this adapter is quite similar to LoHa, but from my point of view it is quite hard to reuse LoHa code for LoKr to be easier to maintain.

Currently the code is working fine, we can train / convert / infer existing LoKr checkpoints from [civitai.com](http://civitai.com).

I've also refactored the script for [SD dreambooth training](https://github.com/huggingface/peft/blob/7526aa224f5fb06d1b5bfe5e94d14f83094f7f06/examples/stable_diffusion/train_dreambooth.py) and incorporated subparsers to simplify training with different adapters (lora, loha, lokr). Here are the results from my sample training which shows that this LoKr implementation is able to train (the output adapter weight is several megabytes in size with requested rank 32):

![Screenshot 2023-10-04 at 14 24 17](https://github.com/huggingface/peft/assets/1431206/fd1c7e30-b012-4883-99d8-d33cde8ea1a9)

I used the following settings for training:

```sh
python train_dreambooth.py \
--pretrained_model_name_or_path=... \
--instance_data_dir=... \
--instance_prompt="AlexanderKovalchuk" \
--output_dir=... \
--seed=42 \
--resolution=512 \
--train_text_encoder \
--train_batch_size=2 \
--max_train_steps=1000 \
--learning_rate=1e-4 \
--num_validation_images=4 \
--validation_steps=50 \
--validation_prompt="AlexanderKovalchuk" \
--logging_dir=./output_logs \
--report_to=tensorboard \
--lr_warmup_steps=300 \
--lr_scheduler=constant_with_warmup \
lokr --unet_r=32 --unet_alpha=32 --te_r=32 --te_alpha=32 --unet_use_effective_conv2d --unet_decompose_both --te_decompose_both
```

kovalexal commented:
We may need to update the docs for LoHa in LoKr in the following PRs, from my point of view LoKr is a bit tricky.

1. At first, LoKr decomposes $`\Delta W`$ into two matrices $`\Delta W_1`$ and $`\Delta W_2`$ so that:
    $`\Delta W = \Delta W_1 \otimes \Delta W_2`$
    $`\Delta W \in \mathbb{R}^{N \times M}, \Delta W_1 \in \mathbb{R}^{N_1 \times M_1}, \Delta W_2 \in \mathbb{R}^{N_2 \times M_2}`$
    $`N = N_1 N_2, M = M_1 M_2, N_1 < N_2, M_1 < M_2`$

    There may exist multiple decompositions and `decompose_factor` controls how this decomposition is performed. If `-1` is passed, it will try to decompose sizes to be close to square root of the dimension, otherwise it will try to find the closes decomposition to the number you've provided. In general, the first matrix is smaller than the second one.

2. Then the resulting matrices $`\Delta W_1`$ and $`\Delta W_2`$ are decomposed using the provided rank with the following logic:
    
    - In general, $`\Delta W_1`$ is not decomposed (because it often has a small size) unless requested with `decompose_both`
    - If the requested rank is lower than the smallest dimension if $`\Delta W_{i}`$, it is decomposed to two matrices $`\Delta W_{ia}`$ and $`\Delta W_{ib}`$, $`\Delta W_{ia} \in \mathbb{R}^{X \times r}`$, $`\Delta W_{ib} \in \mathbb{R}^{r \times Y}`$
    - If the requested rank is bigger than the smallest dimension if $`\Delta W_{i}`$, matrix is not decomposed
    - For Conv2d 3x3 layer $`\Delta W_2`$ we may perform Tucker decomposition (as we do it in LoHa - `use_effective_conv2d`)

kovalexal commented:
@BenjaminBossan I've also spotted several errors in [convert_sd_adapter_to_peft.py](https://github.com/huggingface/peft/blob/7526aa224f5fb06d1b5bfe5e94d14f83094f7f06/examples/stable_diffusion/convert_sd_adapter_to_peft.py) script, so I fixed them in this PR.

kovalexal commented:
@BenjaminBossan, would you have some time to review this PR?

BenjaminBossan commented:
Sorry for the delay, the last few days were very busy. I'll try to give a review today or in the next few days.

BenjaminBossan commented:
I didn't have time for a full review yet (hopefully tomorrow), but did take a quick look. As always from you, it already looks quite excellent. Great attention to detail and thanks also for including tests.

I saw that the `LoKrLayer` does not yet have the changes from #979, could you please update it?

When it comes to code duplication, when it comes to `layer.py`, there are quite a few differences between this and LoHa, so I think there is no reasonable way to abstract here. When it comes to `model.py`, however, we might have a chance. The main differences I could spot where some changed variable names in code/docstrings. Do you think we have a chance of re-using a base model class here?

kovalexal commented:
@BenjaminBossan, hi!

I updated `LoKrLayer` according to #979.

I've also heavily refactored `LoHaModel` and `LoKrModel` - feel free to share your thoughts about the current implementation.

BenjaminBossan commented on src/peft/tuners/lycoris_utils.py:
```diff
@@ -0,0 +1,366 @@
+        ...
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        previous_dtype = x.dtype
```
This may need adjusting, depending on whether #1010 is merged.

BenjaminBossan commented on src/peft/tuners/lycoris_utils.py:
```diff
@@ -0,0 +1,366 @@
+    def _check_target_module_exists(config, key):
+        return check_target_module_exists(config, key)
+
+    def _create_and_replace(
```
For this method and `_create_new_module` below, I wonder if the abstraction would still work once more layer types are added, e.g. bnb layers. Do you think this could be extended easily and in a general fashion or is the abstraction maybe overeager? For instance, I wonder how this base method would need to be changed if we wanted to extend it to LoRA.

kovalexal replied:
TBH this whole abstraction seems a bit overeager, primarily because we still need to process each type of input layer separately in `_create_new_module`. But right now, I don't see, why it won't work for LoRA and bnb layers as well.

https://github.com/huggingface/peft/blob/299de88cfcefd6ea2aea4fb22bf4f0ec124af46e/src/peft/tuners/loha/model.py#L81-L84

BenjaminBossan commented on src/peft/tuners/lycoris_utils.py:
```diff
@@ -0,0 +1,366 @@
+            raise ValueError("Please specify `target_modules` in `peft_config`")
+        return peft_config
+
+    @classmethod
```
I see that this wouldn't work with `staticmethod` as in LoRA right now, but do we need `classmethod`? I think a normal method would work here, using `self.prefix` below.

kovalexal replied:
You are right.

BenjaminBossan commented on tests/test_custom_models.py:
```diff
@@ -144,6 +144,24 @@
+        },
+    ),
+    ("Conv2d 1 LOKR", "Conv2d", LoKrConfig, {"target_modules": ["conv2d"]}),
+    ("Conv2d 2 LOKR", "Conv2d", LoKrConfig, {"target_modules": ["conv2d", "lin0"]}),
```
Can we add test cases for `use_effective_conv2d=True/False`, as well as `decompose_both=True/False`? Would it be possible to test different values of `decompose_factor`?

kovalexal replied:
I've also tried to add similar tests for LoHa and discovered a serious bug #1021

BenjaminBossan reviewed (changes requested):
Finally I got around to giving this a proper review. Thanks for your patience.

This looks already quite good. Big thanks also for working on the abstractions, it's really nice to see how much code duplication can be avoided that way. I wonder if we can extend it to encompass LoRA as well (of course that would be a separate PR).

When it comes to the actual implementation of the LoKr method, I honestly haven't checked the details, so I'd trust you and the pre-existing code here to do the right thing.

Regarding the name of `LyCORISConfig` etc.: I could see that this may be confusing to some, though I understand the choice. From a pure spelling perspective, I wonder, however, if we should switch to `LycorisConfig` etc. This is much easier to type and also more consistent with other parts of PEFT (`LoraConfig`, not `LoRAConfig`). IMHO, it's also easier to read (Lycoris-Layer, not LyCORI-SLayer).

Regarding the update to `train_dreambooth.py`, which can now also do LoHa, does it mean that `train_dreambooth_loha.py` can be deleted?

kovalexal commented:
@BenjaminBossan, thank you for giving a review on this PR!

> I wonder if we can extend it to encompass LoRA as well (of course that would be a separate PR).

I suppose that there is no problem in incorporating LoRA into this config, we can do it in a separate PR (taking into account, that we'll have to add/modify additional dropout strategies from LyCORIS).

> Regarding the name of LyCORISConfig etc.: I could see that this may be confusing to some, though I understand the choice.

No problem, I agree with that. Don't you mind if I also refactor LoHa / LoKr using the same naming strategy (in this PR)? LoHa adapter was not released yet, so it won't be a problem as far as I think.

> Regarding the update to train_dreambooth.py, which can now also do LoHa, does it mean that train_dreambooth_loha.py can be deleted?

Definitely, I'll remove it in the following commits.

kovalexal commented on examples/stable_diffusion/convert_sd_adapter_to_peft.py:
```diff
@@ -35,7 +44,7 @@ def peft_state_dict(self) -> Dict[str, torch.Tensor]:
         return {
             f"base_model.model{self.peft_key}.lora_A.weight": self.lora_A,
-            f"base_model.model.{self.peft_key}.lora_B.weight": self.lora_A,
+            f"base_model.model.{self.peft_key}.lora_B.weight": self.lora_B,
```
We'll definitely need to come up with regression tests for this script one day to prevent those types of errors.

kovalexal commented on src/peft/tuners/loha/layer.py:
```diff
@@ -14,57 +14,46 @@
-
-    def create_loha_parameters(self, adapter_name: str, r: int, shape: Tuple[int, ...]):
+    def _available_adapters(self) -> Iterable[str]:
+        return set(
```
Nice!

kovalexal commented on examples/stable_diffusion/train_dreambooth.py:
```diff
@@ -85,6 +85,80 @@ def import_model_class_from_model_name_or_path(pretrained_model_name_or_path: st
+    return config
+
+
+def create_te_adapter_config(args: argparse.Namespace) -> Union[LoraConfig, LoHaConfig, LoKrConfig]:
```
Ok. Are you comfortable with shortened naming convention for CLI parameters? I find it cumbersome and IMO it's a bit annoying to type several parameters like `text_encoder_decompose_factor`, etc.

BenjaminBossan commented:
> No problem, I agree with that. Don't you mind if I also refactor LoHa / LoKr using the same naming strategy (in this PR)? LoHa adapter was not released yet, so it won't be a problem as far as I think.

I guess your suggestion is to move to `Loha` and `Lokr` to be consistent with `Loha`, right?  think for those two, it's fine either way (for instance, we also use `IA3` and not `Ia3`, so we're not perfectly consistent).

For LoHa, ideally, it would be a separate PR. The reason is that we aim at having a release soon-ish once we return from the PT conference. That's probably too soon for this PR to be finished, so we may end up with `LoHa` being used in the release version. A separate PR would be much quicker to be merged.

kovalexal commented:
> For LoHa, ideally, it would be a separate PR. A separate PR would be much quicker to be merged.

Ok, I'll do my best to create with PR for `LoHa` -> `Loha` transition ASAP.

kovalexal commented on src/peft/tuners/lokr/layer.py:
```diff
@@ -0,0 +1,361 @@
+
+def factorization(dimension: int, factor: int = -1) -> Tuple[int, int]:
+    """
+    return a tuple of two value of input dimension decomposed by the number closest to factor second value is higher or
```
Sorry, this was copy-pasted. I'll rewrite this docstring and provide doctest-style example.

kovalexal replied:
Done.

kovalexal commented on examples/stable_diffusion/convert_sd_adapter_to_peft.py:
```diff
@@ -73,7 +82,49 @@ def peft_state_dict(self) -> Dict[str, torch.Tensor]:
+    lokr_t2: Optional[torch.Tensor] = None
+
+    def peft_state_dict(self) -> Dict[str, torch.Tensor]:
+        if (self.lokr_w1 is None and self.lokr_w1_a is None and self.lokr_w1_b is None) or (
+            self.lokr_w2 is None and self.lokr_w2_a is None and self.lokr_w2_b is None
+        ):
```
You are right, that's not sufficient.

kovalexal commented on examples/stable_diffusion/convert_sd_adapter_to_peft.py:
```diff
@@ -73,7 +82,49 @@ def peft_state_dict(self) -> Dict[str, torch.Tensor]:
+        return state_dict
+
+
+def construct_peft_loraconfig(info: Dict[str, LoRAInfo], **kwargs) -> LoraConfig:
```
Yes, it's okay. This is required to provide the same interface between LoRA / LoHa / LoKr in case they take different sets of parameters.

For example, `decompose_factor` is a global parameter for a module in the case of LoKr, so it needs to be passed somewhere.

kovalexal commented:
@BenjaminBossan, I addressed most of the comments you've left.

> The reason is that we aim at having a release soon-ish once we return from the PT conference. That's probably too soon for this PR to be finished

I have one idea, but I am not sure, whether you will find it useful and appropriate, or not. I had a thought that we may showcase that PEFT is able to load most of the adapters for Stable Diffusion from civitai.com through a blog post on the Hugging Face blog (as far as I know, PEFT is the only way to achieve this with Hugging Face ecosystem right now). We may collaborate and co-author and I am sure that it would help to attract more users of SD/SDXL to your library.

In case you see it useful as well it probably would be great to do it after the LoKr adapter finally gets into release.

BenjaminBossan commented:
> I have one idea, but I am not sure, whether you will find it useful and appropriate, or not. I had a thought that we may showcase that PEFT is able to load most of the adapters for Stable Diffusion from civitai.com through a blog post on the Hugging Face blog (as far as I know, PEFT is the only way to achieve this with Hugging Face ecosystem right now). We may collaborate and co-author and I am sure that it would help to attract more users of SD/SDXL to your library.

I'm personally not an expert on SD and the ecosystem, but I'm sure we can manage something. Let me discuss this with my colleagues.

Meanwhile, could you please fix the merge conflicts and then I can give a (hopefully) final review.

kovalexal commented:
> Meanwhile, could you please fix the merge conflicts and then I can give a (hopefully) final review.

Sure, I fixed all the merge conflicts.

BenjaminBossan commented on src/peft/tuners/lycoris_utils.py:
```diff
@@ -0,0 +1,365 @@
+        ...
+
+
+class LyCORISTuner(BaseTuner):
```
Could you please rename this class to `LycorisTuner` for consistency?

BenjaminBossan commented on src/peft/tuners/lokr/layer.py:
```diff
@@ -0,0 +1,371 @@
+        """Internal function to create lokr adapter
+
+        Args:
+            shape (`Tuple[int, ...]`): Shape of weights to produce
```
`shape` is not an argument of this function, so it can be removed. On the other hand, some arguments at the end are missing.

kovalexal replied:
Thank you, I missed it during refactoring.

BenjaminBossan commented on src/peft/tuners/lycoris_utils.py:
```diff
@@ -0,0 +1,365 @@
+    r"""
+    A base config for LyCORIS like adapters
+    """
+    rank_pattern: Optional[dict] = field(
```
In `LohaConfig` and `LoKrConfig`, the arguments `rank_pattern` and `alpha_pattern` are also defined. Does it make sense to have them in parent class and child class?

kovalexal replied:
You are right, I just wanted to leave them explicit in case anyone would read the code, but from the other point it may lead to some errors in case of refactoring or some other modifications, so let's remove them from child classes.

BenjaminBossan commented on src/peft/tuners/lycoris_utils.py:
```diff
@@ -0,0 +1,365 @@
+
+    @property
+    def merged(self) -> bool:
+        return bool(self.merged_adapters)
```
Can we move this `property` to `BaseTunerLayer`? Then we can remove it here, as well as from `IA3Layer` and `LoraLayer`.

kovalexal replied:
Sure, seems a good point of improvement!

BenjaminBossan commented on src/peft/tuners/lycoris_utils.py:
```diff
@@ -0,0 +1,365 @@
+                self.weight.data += self.get_delta_weight(active_adapter)
+                self.merged_adapters.append(active_adapter)
+
+    def reset_adapter_parameters(self, adapter_name: str):
```
Is this really needed as an abstract method? The abstract class does not require it.

kovalexal replied:
Yes, it should be an abstract method, so all child adapters need to implement them in their own way.
I missed to explicitly state that these methods should be abstract, so thank you for bringing it up!

BenjaminBossan commented on src/peft/tuners/lycoris_utils.py:
```diff
@@ -0,0 +1,365 @@
+    def reset_adapter_parameters(self, adapter_name: str):
+        ...
+
+    def scale_layer(self, scale_factor: float) -> None:
```
We made some small changes to `scale_layer`, `unscale_layer`, and added a new method `set_scale` in these PRs: #1029, #1034. Could you please update your code accordingly?

BenjaminBossan commented on src/peft/tuners/lycoris_utils.py:
```diff
@@ -0,0 +1,365 @@
+            r = self.r[active_adapter]
+            self.scaling[active_adapter] = alpha / r
+
+    def update_layer(self, adapter_name: str, r: int, alpha: float, **kwargs):
```
Same as above: Is this really needed as an abstract method? The abstract class does not require it.

kovalexal replied:
Answered above.

BenjaminBossan commented on tests/test_custom_models.py:
```diff
@@ -445,6 +489,10 @@ def test_only_params_are_updated(self, test_name, model_id, config_cls, config_k
         params_after = dict(model.named_parameters())
         self.assertEqual(params_before.keys(), params_after.keys())
 
+        if isinstance(model, ModelConv2D):
```
Probably some debug code, please remove.

kovalexal replied:
You are right, sorry for that.

BenjaminBossan reviewed (changes requested):
So I did another review and it seems like we're almost done. I have some minor comments left, please take a look.

BenjaminBossan reviewed (approved):
Great work, Alex, this looks stellar to me. Thanks for all the effort you put into this.

From my point of view, this is ready, I'll check with colleagues if we want to have a second review just to be sure.

I took a closer look at test coverage and found some small gaps, but they're not blockers for me. Up to you if you want to work on those:

1. We can improve test coverage slightly by adding a LoKr config with rank dropout, e.g.:

`("Vanilla MLP 7 LOKR", "MLP", LoKrConfig, {"target_modules": "lin0", "rank_dropout": 0.5}),`

2. There is currently no code path that leads to `use_w1=False` or to `use_w2=False` with conv2d. Honestly, not sure what needs to be done to get those.

BenjaminBossan commented:
Found a tiny are for improvement: In the docstring of `check_target_module_exists` in `tuners_utils.py`, the type for config is given as `LoraConfig | LoHaConfig`, it can be updated to `LoraConfig | LycorisConfig`.

kovalexal commented:
@BenjaminBossan, I updated the docstring of `check_target_module_exists` and also added several tests to cover `rank_dropout` functionality for LoHa & LoKr + path that leads to `use_w1=False` and `use_w2=False`.

BenjaminBossan commented:
> I updated the docstring of `check_target_module_exists` and also added several tests to cover `rank_dropout` functionality for LoHa & LoKr + path that leads to `use_w1=False` and `use_w2=False`.

Great, thanks a lot.

BenjaminBossan commented:
@kovalexal I just noticed that LoHa and LoKr are missing `delete_adapter`, or am I missing it? If true, it can be added in a later PR, just wanted to confirm.

kovalexal commented:
@BenjaminBossan Yes, it slipped under my radar (added it in the recent commit), thank you for noticing!

BenjaminBossan commented:
Oh, noticed a small error in your last commit:

> [key for key, _ in self.model.named_modules() if **"lora"** not in key]

kovalexal commented:
@BenjaminBossan sorry, fixed that one.

younesbelkada commented on src/peft/tuners/ia3/layer.py:
```diff
@@ -43,10 +43,6 @@ def __init__(
         self.out_features = out_features
         self.is_feedforward = is_feedforward
 
-    @property
```
Why this has been removed?

younesbelkada commented on src/peft/tuners/tuners_utils.py:
```diff
@@ -281,12 +281,19 @@ class BaseTunerLayer(ABC):
     def unmerge(self, *args) -> None:
         raise NotImplementedError
 
+    @property
```
I see, because you moved it here

younesbelkada reviewed (approved):
Thanks!
